lib/nostrum/struct/channel.ex

```diff
@@ -0,0 +1,266 @@
+  @moduledoc ~S"""
+  Struct representing a Discord guild channel.
+
+  A `Nostrum.Struct.Channel` represents all 5 types of channels. Each
+  channel has a field `:type` with any of the following values:
+
+    * `0` - GUILD_TEXT
+    * `1` - DM
+    * `2` - GUILD_VOICE
+    * `3` - GROUP_DM
+    * `4` - GUILD_CATEGORY
+
+  More information can be found on the
+  [Discord API Channel Documentation](https://discordapp.com/developers/docs/resources/channel#channel-object).
+
+  ## Mentioning Channels in Messages
+
+  A `Nostrum.Struct.Channel` can be mentioned in message content using the `String.Chars`
+  protocol or `mention/1`.
+
+  ```Elixir
+  channel = %Nostrum.Struct.Channel{id: 381889573426429952}
+  Nostrum.Api.create_message!(184046599834435585, "#{channel}")
+  %Nostrum.Struct.Message{content: "<#381889573426429952>"}
+
+  channel = %Nostrum.Struct.Channel{id: 280085880452939778}
+  Nostrum.Api.create_message!(280085880452939778, "#{Nostrum.Struct.Channel.mention(channel)}")
+  %Nostrum.Struct.Message{content: "<#280085880452939778>"}
+  ```
+  """
+
+  alias Nostrum.Struct.Overwrite
+  alias Nostrum.Struct.Snowflake
+  alias Nostrum.Struct.User
+  alias Nostrum.Util
+
+  defstruct [
+    :id,
+    :type,
+    :guild_id,
+    :position,
+    :permission_overwrites,
+    :name,
+    :topic,
+    :nsfw,
+    :last_message_id,
+    :bitrate,
+    :user_limit,
+    :recipients,
+    :icon,
+    :owner_id,
+    :application_id,
+    :parent_id,
+    :last_pin_timestamp
+  ]
+
+  defimpl String.Chars do
+    def to_string(channel), do: @for.mention(channel)
+  end
+
+  @typedoc "The channel's id"
+  @type id :: Snowflake.t()
+
+  @typedoc "The id of the channel's guild"
+  @type guild_id :: Snowflake.t()
+
+  @typedoc "The ordered position of the channel"
+  @type position :: integer
+
+  @typedoc "The list of overwrites"
+  @type permission_overwrites :: [Overwrite.t()]
+
+  @typedoc "The name of the channel"
+  @type name :: String.t()
+
+  @typedoc "Current channel topic"
+  @type topic :: String.t()
+
+  @typedoc "If the channel is nsfw"
+  @type nsfw :: boolean
+
+  @typedoc "Id of the last message sent"
+  @type last_message_id :: Snowflake.t() | nil
+
+  @typedoc "The bitrate of the voice channel"
+  @type bitrate :: integer
+
+  @typedoc "The user limit of the voice channel"
+  @type user_limit :: integer
+
+  @typedoc "The recipients of the DM"
+  @type recipients :: [User.t()]
+
+  @typedoc "The icon hash of the channel"
+  @type icon :: String.t() | nil
+
+  @typedoc "The id of the DM creator"
+  @type owner_id :: Snowflake.t()
+
+  @typedoc "The application id of the group DM creator if it is bot-created"
+  @type application_id :: Snowflake.t() | nil
+
+  @typedoc "The id of the parent category for a channel"
+  @type parent_id :: Snowflake.t() | nil
+
+  @typedoc "When the last pinned message was pinned"
+  @type last_pin_timestamp :: String.t() | nil
+
+  @type guild_text_channel :: %__MODULE__{
```

Comment:
Do you think it would be useful to include documentation for these types and the type "groups" down below?